_posts/2020-05-22-validate-spring-boot-configuration-parameters-at-startup.md

```diff
@@ -0,0 +1,225 @@
+
+### Nested Configuration Objects
+For some properties, it would be meaningful to bind them into a nested object.
+
+So, let's create `ReportProperties` to group the properties related to `report` key:
+
+```java
+class ReportProperties {
+
+  private Boolean sendEmails = Boolean.FALSE;
+
+  private ReportType type = ReportType.HTML;
+
+  @Min(value = 7)
+  @Max(value = 30)
+  private Integer intervalInDays;
+
+  @Email
+  private String emailAddress;
+
+  // ...
+}
+```
+
+Next, we refactor our `AppProperties` to include our nested object `ReportProperties`:
+
+```java
+@Validated
+@ConfigurationProperties(prefix="app.properties")
+class AppProperties {
+
+  @NotEmpty
+  private String name;
+
+  @Valid
+  private ReportProperties report;
+
+  // ...
+}
+```
+
+Besides, we should also pay attention to put `@Valid` annotation on our nested class field `report`.
+
+Thus, Spring can validate the inner properties included in nested objects.
+
+### Using Bean Methods
+Another way of triggering validation, we can use the Spring bean methods. This is particularly useful when we want to **bind properties to components defined in third-party libraries or maintained in separate jar files**:
+```java
+@Bean
+@Validated
+@ConfigurationProperties(prefix = "app.third-party.properties")
+public ThirdPartyComponentProperties thirdPartyComponentProperties() {
+  return ThirdPartyComponentProperties();
+}
+```
+
+### Using a Custom `Validator`
+Even though bean validation provides a declarative approach to validate our objects in a reusable way, sometimes we need more to customize our validation logic in purpose.
+
+In that case, **Spring has an independent `Validator` mechanism to allow dynamic input validation**.
+
+Let's make our validation further, to check the `report.email-address` can only be an address under a specific domain like `@analysisapp.com`:
+```java
+class ReportEmailAddressValidator implements Validator {
+
+  private static final String EMAIL_DOMAIN = "@analysisapp.com";
+
+  public boolean supports(Class clazz) {
+    return ReportProperties.class.isAssignableFrom(clazz);
+  }
+
+  public void validate(Object target, Errors errors) {
+
+    ValidationUtils.rejectIfEmptyOrWhitespace(errors,
+        "emailAddress", "field.required");
+
+    ReportProperties reportProperties = (ReportProperties) target;
+    if (reportProperties.getEmailAddress().endsWith(EMAIL_DOMAIN)) {
+      errors.rejectValue("emailAddress", "field.domain.required",
+          new Object[]{EMAIL_DOMAIN},
+          "The email address must contain [" + EMAIL_DOMAIN + "] domain.");
+    }
+
+  }
+}
+```
+Then, we need to register our custom Spring validator with a specific definition of `configurationPropertiesValidator`:
+```java
+@Bean
+public static ReportEmailAddressValidator configurationPropertiesValidator(){
```

Comment:
What I don't understand from reading the text is how does the validator know which fields it should validate?

Ah, it's the special method name `configurationPropertiesValidator`? But still, how does it know that it should only validate the email field?

Reply:
Actually it doesn't validate the field but the whole object. It just checks with the `boolean supports(Class clazz)` method if the object is eligible to validate and then executes the custom logic we implemented inside `validate` method. We throw error messages with the field names as we like in the implementation.

Yes, the special bean definition `configurationPropertiesValidator` configures the `Validator` to work right after the bean validation for configuration properties, that's good. 

However, what I also couldn't figure out what if we want to define more than one `Validator` for configuration properties? Let me have a look at that also.

Reply:
Found a solution in: [https://stackoverflow.com/questions/33838978/multiple-configurationproperties-validator-beans-in-spring-environment](https://stackoverflow.com/questions/33838978/multiple-configurationproperties-validator-beans-in-spring-environment)

I think I'm going to add these to the article also.

Reply:
I've added an example code about how to define custom `Validator` more than one for configuration properties. Also, updated the code examples